MIPS porting (preliminary)

Harry-Chen opened the pull request:
* Port rCore to `mipsel` architecture
* Runs on qemu malta machine with uart and PCI (enumeration only) support
* Refactor some drivers
* Successfully runs most user apps

jiegec commented on bootloader/src/arch/mipsel/boot.S:
```diff
@@ -0,0 +1,20 @@
+.globl _start
+
+_start:
+    # read cpu affinity, start core 0, halt the rest
```
Waiting for multicore support

Harry-Chen replied:
not now

Harry-Chen replied:
will add multicore support, maybe later this month

jiegec commented on kernel/src/arch/mipsel/atomic.patch:
```diff
@@ -0,0 +1,72 @@
+--- atomic_backup.rs	2018-10-06 19:59:14.000000000 +0800
```
maybe not needed anymore in misp?

Harry-Chen replied:
Will delete

jiegec commented on kernel/src/arch/mipsel/board/malta/mod.rs:
```diff
@@ -0,0 +1,54 @@
+/// Initialize other board drivers
+pub fn init_driver() {
+    // TODO: add possibly more drivers
+    vga::init(0xbbe00000, 0xb2050000, 800, 600);
```
can we find these address in dtb?

Harry-Chen replied:
NO. VGA is needed by framebuffer driver and video console, which initializes BEFORE the driver framework.

Also this is a board specific file, I don't see any problem with hardcoding addresses.

Harry-Chen replied:
Maybe a comprehensive refactor of the driver framework can solve this.

jiegec commented on kernel/src/arch/mipsel/boot/entry.S:
```diff
@@ -0,0 +1,40 @@
+_start:
+    # setup stack and gp
+	la    sp, bootstacktop
+    la    gp, _gp
```
Mixed tab and space?

Harry-Chen replied:
Will fix

jiegec commented:
LGTM

Harry-Chen commented:
RV32 and RV64 tests failed to pass. However, it does not seems to be our problem. 

Fixed some minor problems pointed out by jiegec and added mipsel to CI.

wangrunji0408 commented:
I'm reviewing...
Orz first for your great contribution.

A personal suggestion: 
It's unfriendly to macOS developers now. 😂
Is there a binary release of `mipsel-linux-musln32-gcc` on macOS?
Or could we get rid of the gcc toolchain for kernel?
It seems that only assembly code depends on it.
I think `.macro` or `.equ` in asm may achieve the same effect.

And it's better to run `cargo fmt` before the final merge.

Harry-Chen commented:
There seems no .macro support in MIPS assembly, and I could not figure out why. But to format the code is a good suggestion.

Harry-Chen commented:
But I can change to any version of `gcc`, since we only use it for preprocessing.

wangrunji0408 commented:
A crazy idea: preprocess the asm using `build.rs`

Harry-Chen commented:
> A crazy idea: preprocess the asm using `build.rs`

Might work. But I have now changed to `gcc`, which I believe adds no difficulty to any developer

wangrunji0408 commented on kernel/src/arch/mipsel/compiler_rt.rs:
```diff
@@ -0,0 +1,8 @@
+//! [atomic](http://llvm.org/docs/Atomics.html#libcalls-atomic)
+
+#[no_mangle]
+pub extern "C" fn abort() {
```
This can be removed on mips.

wangrunji0408 commented on kernel/src/arch/mipsel/context.rs:
```diff
@@ -0,0 +1,273 @@
+    pub sp: usize,
+    pub fp: usize,
+    pub ra: usize,
+    /// Reserve space for hartid
```
No longer needed since we can get core id from `cp0.ebase`
